registrations/serializers.py

```diff
@@ -35,6 +35,32 @@ def _validate_registration_enrolment_times(registration: Registration) -> None:
         raise ConflictException(_("Enrolment is already closed."))
 
 
+def _validate_registration_capacity(
+    registration: Registration, signups_count: int
+) -> tuple[int, int]:
+    attendee_capacity = registration.maximum_attendee_capacity
+    waiting_list_capacity = registration.waiting_list_capacity
+    add_as_attending = add_as_waitlisted = signups_count
+
+    if attendee_capacity is not None:
+        already_attending = SignUp.objects.filter(
+            registration=registration, attendee_status=SignUp.AttendeeStatus.ATTENDING
+        ).count()
+        add_as_attending = attendee_capacity - already_attending
+
+    if waiting_list_capacity is not None:
+        already_waitlisted = SignUp.objects.filter(
+            registration=registration,
+            attendee_status=SignUp.AttendeeStatus.WAITING_LIST,
+        ).count()
+        add_as_waitlisted = waiting_list_capacity - already_waitlisted
+
+    if not (add_as_attending or add_as_waitlisted):
+        raise DRFPermissionDenied(_("The waiting list is already full"))
```

Comment:
Two thoughts

1) I suppose `add_as_attending` and `add_as_waitlisted` can, at least in theory, be negative numbers if the number of attendees/waitlisters exceeds the set capacity (can capacity be modified to be less than the number of existing items?). A negative number is truthy, so it would sneak past this if case. Not a big problem, as the create() will not do anything, except maybe result in error 500 instead of 403. 

2) I wouldn't mind keeping this where it was before, at the end of the view. Now the view has `if add_as_attending` and `elif add_as_waitlisted`, but it is not immediately obvious what the "else" behaviour is until you take a look at this function. If that makes sense?

Reply:
I think this is a very good suggestion and the negative numbers being truthy slipped my mind completely. Now improvements have been made based on these points.